### ISSUE TITLE
Add support for more complex signatures to ufunc

### ISSUE THREAD

**Issue by cosama** [numpy/numpy#18151]
## Feature

This feature request is a follow up of https://github.com/numba/numba/issues/6617.

I am not familiar enough with `ufunc` to understand the feasibility or even if this is already possible when implementing "pure" `ufunc`s. However, a similar use case is the `numpy.vectorize` class which takes a `signature` argument.

What I would like to do is to perform simple arithmetic operations in these signatures. For example a use case that comes up often in my work with histograms is the need for a signature of the form:

`(n + 1)->(n)`.

The input array contains edges, thus is one element longer than the output array with the bin values.

In general I think it would be useful to support some very simple arithmetic operations in the signature (i.e. addition, multiplication, etc) of given inputs. For example I can imagine this one also to be quite useful:

`(n),(m)->(m*n)`

My understanding is, that if this would be implemented in NumPy a package such as `numba` could directly use this and allow for more flexibility in their own `ufunc` translator as well.

Thanks.

**Comment by seberg** [comment]
Hmmm, this is an interesting problem. If we go down this road, I am wondering if this should actually go one step further:
Instead of allowing a more complex mini-language, expose a "transform" hook/function that returns the shape and axes mapping that we need to set up the iteration.  (I have to carefully think how it fits with [NEP 43](https://numpy.org/neps/nep-0043-extensible-ufuncs.html), but I doubt it will matter much. Signatures will probably be a bit annoying either way with NEP 42 and I just have to pick some decent solution.)

The signature is parsed into a simplified description here: https://github.com/numpy/numpy/blob/c946dbaed8737af68fee8de44cbde7169a8ec75c/numpy/core/src/umath/ufunc_object.c#L617

And this is used to construct the correct iterator (and fill in the additional fixed sizes in this, pretty large, code block (not quite all of this is the core dimensions): https://github.com/numpy/numpy/blob/c946dbaed8737af68fee8de44cbde7169a8ec75c/numpy/core/src/umath/ufunc_object.c#L2559-L2756

Long story short: I think we could do this, but it needs a good proposal (either extending the string syntax, or allowing to "hook" with a function) and some thought proposal. A very brief NEP based on the initial one at https://numpy.org/neps/nep-0020-gufunc-signature-enhancement.html may also make sense, if it is not very clear how to do it best.

I expect the main issue is that I doubt anyone from the core team can champion this at the time (happy to help though, @mhvk may have a opinions/thoughts on this), at least unless this is widely important/useful.  But maybe we get lucky :).

**Comment by mhvk** [comment]
I like the idea very much! But implementing it directly will not be super easy - as it is, the code is not really set up to deal with relations between dimensions. It would also need some strong use cases (I'm still a bit sad my `n|1` idea for broadcastable core dimensions ended up not accepted, and that was mostly because the use case was considered too weak). 

I do really like @seberg's suggestion of having a possible call-back, as it would open up also completely different ideas (I vaguely feel I've lamented not having it, but am not sure what it was about any more...). Obviously, it would need something with a well-described API, so that packages like numba can use it (a good interface to the present parser would be a nice by-product...).

**Comment by cosama** [comment]
@seberg @mhvk Thanks for the fast reply and all the great links. I am very happy to see that there is actually some interest in this idea. I really like the idea of a "transform" hock/function, that would for sure address my needs. 

I suspect it indeed might be hard to argue for this though as direct use cases seem limited (I really only can think of the histogram one right now). I am happy to do a bit more research though and help work this out if there is something I can do with my "limited" knowledge of the `ufunc` internals.

**Comment by WarrenWeckesser** [comment]
I just (re)discovered this enhancement request.  Last year I wrote up some brief notes for a basically the same idea over in https://github.com/WarrenWeckesser/numpy-notes/blob/master/enhancements/gufunc-output-size-exressions.md
